### PR TITLE
[slack-usergroups] Update Slack usergroups only when there is a diff

### PR DESCRIPTION
**Pull request by steveteahan**
This change saves a call to usergroups.update and usergroups.users.update
per usergroup for cases where there isn't a diff. Both of these APIs are
Tier 2 Slack API calls only allowing ~20 calls per minute. This can save
hundreds of calls per minute assuming a modest number of Slack usergroups.

Additionally, performed some refactoring to log inline with Slack API
calls and improved the data structures used to represent state for
simpler lookups.

### CODE REVIEW

**Comment by steveteahan** [comment]
app-sre specific testing is also being tracked here: https://issues.redhat.com/browse/APPSRE-3643

File: reconcile/slack_cluster_usergroups.py [hunk 1]
```diff
@@ -75,42 +85,56 @@ def get_desired_state(slack):
...
 
     return current_state
 
 
 def run(dry_run):
     slack = init_slack_workspace(QONTRACT_INTEGRATION)
     desired_state = get_desired_state(slack)
-    usergroups = [d['usergroup'] for d in desired_state]
+    usergroups = []
+    for _, workspace_state in desired_state.items():
```
Comment (Piojo) [inline reconcile/slack_cluster_usergroups.py:133]: This is easier as
```suggestion
    for workspace_state in desired_state.values()
```

File: reconcile/slack_usergroups.py [hunk 2]
```diff
@@ -239,11 +249,24 @@ def get_slack_usernames_from_owners(owners_from_repo, users, usergroup):
 
 
 def get_desired_state(slack_map, pagerduty_map):
+    """
+    Get the desired state of Slack usergroups.
+
+    :param slack_map: Slack data from app-interface
```
Comment (Piojo) [inline reconcile/slack_usergroups.py:255]: Can't wait for our discussion on type hints...

File: reconcile/slack_usergroups.py [hunk 3]
```diff
@@ -285,97 +308,173 @@ def get_desired_state(slack_map, pagerduty_map):
...
+    users = list(desired_ug_state['users'].keys())
+
+    current_users = set(current_ug_state.get('users', {}).values())
+    desired_users = set(desired_ug_state['users'].values())
+
+    for user in desired_users - current_users:
+        logging.info(['add_user_to_usergroup',
+                      workspace, usergroup, user])
+
+    for user in current_users - desired_users:
```
[the remark this answers is not in the document]
Comment (Piojo) [reply reconcile/slack_usergroups.py:363]: (sure, keep it in that tuple format if you wish)

**Review by Piojo — APPROVED** [review]
Overall looks great. Do the nits at your discretion.

File: reconcile/slack_usergroups.py [hunk 4]
```diff
@@ -285,97 +308,173 @@ def get_desired_state(slack_map, pagerduty_map):
...
 
+    workspace = desired_ug_state['workspace']
+    usergroup = desired_ug_state['usergroup']
+    ugid = desired_ug_state['usergroup_id']
+    users = list(desired_ug_state['users'].keys())
+
+    current_users = set(current_ug_state.get('users', {}).values())
+    desired_users = set(desired_ug_state['users'].values())
+
+    for user in desired_users - current_users:
```
[the remark this answers is not in the document]
Comment (steveteahan) [reply reconcile/slack_usergroups.py:359]: I'm mostly just following the convention of what existed in `print_diff` and other integrations that I've seen.  This is probably part of a bigger discussion about having documented conventions for errors handling in this project like when to use `RunnerException`, what should be logged at `INFO` (and in what format), etc.  Maybe something to discuss further in a future PR that could outline some of these practices?

File: reconcile/slack_cluster_usergroups.py (hunk 1 above)
Comment (steveteahan) [reply reconcile/slack_cluster_usergroups.py:133]: That's a fair point.  I'll consider that pattern in the future when I'm throwing away the other value.  Thanks!

File: reconcile/slack_cluster_usergroups.py [hunk 5]
```diff
@@ -75,42 +85,56 @@ def get_desired_state(slack):
...
+        for usergroup, _ in workspace_state.items():
```
[the remark this answers is not in the document]
Comment (steveteahan) [reply reconcile/slack_cluster_usergroups.py:134]: As with the other, I'll keep this in mind for any future cases.  Thanks!

File: reconcile/slack_usergroups.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (steveteahan) [reply reconcile/slack_usergroups.py:359]: I started a thread on Slack to ask if conventions exist and whether there needs to be any documentation.  For now I'll stick with the existing convention, but I don't disagree with what you're suggesting.

**Comment by steveteahan** [comment]
Thank you for the review, @Piojo!

**Pull request merged**